torch/_higher_order_ops/associative_scan.py

```diff
@@ -28,19 +33,98 @@
     rhs = pytree.tree_unflatten(args[num_leaves:], spec)
     combined = combine_fn(lhs, rhs)
     combined_leaves = pytree.tree_leaves(combined)
-    assert num_leaves == len(combined_leaves)
+    if num_leaves != len(combined_leaves):
+        raise ValueError(
+            "The number of levaes of the inputs need to be identical to the number of leaves of the scan output"
+        )
     return combined_leaves
 
 
+def check_args(combine_fn, leaves, tree, dim):
+    if not callable(combine_fn):
+        raise ValueError("combine_fn must be a callable, but got {combine_fn}")
+    if not isinstance(dim, int):
+        raise ValueError("dim must be an int, but got {type(dim)}")
+
+    if len(leaves) < 1:
+        raise ValueError("expected at least 1 input leaf")
+
+    if not all(isinstance(x, torch.Tensor) for x in leaves):
+        raise ValueError("input leaves must be a Tensor")
+    shape = leaves[0].shape
+    ndim = len(shape)
+    dim = utils.canonicalize_dim(ndim, dim)
+
+    for x in leaves[1:]:
+        if x.shape != shape:
+            raise ValueError("All input tensors must have the same shape")
+
+    out = combine_fn(
+        pytree.tree_unflatten(leaves, tree),
+        pytree.tree_unflatten(leaves, tree),
+    )
+
+    out_leaves, tree_out = pytree.tree_flatten(out)
+    if tree.num_nodes != tree_out.num_nodes or any(
+        o.shape != i.shape or o.dtype != i.dtype or o.device != i.device
+        for o, i in zip(out_leaves, leaves)
+    ):
+        raise ValueError(
+            "The pytree of the output of the operator needs to match the input pytree"
+        )
+
+
+def _interleave(a, b, dim):
+    # https://stackoverflow.com/questions/60869537/how-can-i-interleave-5-pytorch-tensors
+    if b_trunc := (a.shape[dim] == b.shape[dim] + 1):
+        pad = [0, 0] * b.ndim
+        pad[
+            (b.ndim - dim - 1) * 2 + 1
+        ] = 1  # +1=always end of dim, pad-order is reversed so start is at end
+        b = torch.nn.functional.pad(b, pad)
+
+    stacked = torch.stack([a, b], dim=dim + 1)
+    interleaved = torch.flatten(stacked, start_dim=dim, end_dim=dim + 1)
+    if b_trunc:
+        # TODO: find torch alternative for slice_along dim for torch.jit.script to work
+        interleaved = interleaved[
+            slice_along_axis(0, b.shape[dim] + a.shape[dim] - 1, dim=dim)
+        ]
+    return interleaved
+
+
+def safe_map(f, *args):
+    args = list(map(list, args))
+    n = len(args[0])
+    for arg in args[1:]:
+        if len(arg) != n:
+            raise ValueError("length mismatch: {list(map(len, args))}")
+
+    def nf(a):
+        return f(*a)
+
+    return list(map(nf, zip(*args)))
+
+
+def slice_along_axis(start, end, stride=None, dim=0):
+    return (slice(None),) * dim + (slice(start, end, stride),)
+
+
 def associative_scan(
     combine_fn: Callable[[pytree.PyTree, pytree.PyTree], pytree.PyTree],
     input: pytree.PyTree,
     dim: int,
+    reverse: bool = False,
+    generic_scan: bool = False,
```

Comment:
Should we name them as combine_mode= "pointwise"/"generic"/"auto", where "auto" basically means whenever pointwise fails, fallback to generic. In this pr, I think we should focus adding `pointwise`.